### PR TITLE
Add statistics and logging to signup and signin processes #416

### PR DESCRIPTION
**Pull request by simison**
Adds new `signup` stat with a tag:
- `status: success|failure`

Adds new `signin` stat with a tags:
- `status: success|failure`
- `reason: wrong-credentials|suspended|other`

### CODE REVIEW

**Comment by simison** [comment]
@chmac @mrkvon I had some time today in the bus so I added these stats to log failed/successful signups and logins. Should give some indication of activity as well.

@mrkvon am I using the stats module correctly here?

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 1]
```diff
@@ -102,21 +99,53 @@ exports.signup = function(req, res) {
...
+        user = profileHandler.sanitizeProfile(user);
+
+        done(err, user);
       });
     }
 
-  ], function(err) {
+  ], function(err, user) {
+
+    var statsObject = {
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: The `statsObject` looks good. `time`, `meta` and `tags` fields are optional, so `meta` can be removed and `time` depends on preference (time now vs. time of arrival to stathat)

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 2]
```diff
@@ -125,30 +154,89 @@ exports.signup = function(req, res) {
  * Signin after passport authentication
  */
 exports.signin = function(req, res, next) {
+
+  var statsObject = {
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: The same as line 111+.

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 3]
```diff
@@ -125,30 +154,89 @@ exports.signup = function(req, res) {
...
-      return res.status(400).send(info);
+      // Log the failure to signin
+      log('error', 'User signin failed. #3tfgbg-1', {
+        reason: 'Wrong credentials',
+        error: err || null
+      });
+
+      // Send signin failure to stats servers
+      statsObject.tags.status = 'failed';
+      statsObject.tags.reason = 'wrong-credentials';
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: In stathat, `tags.status` and `tags.reason` will currently lead to statistics with names:

- `signin.count`
- `signin.count.status.[failed|success]`
- `signin.count.reason.[wrong-credentials|suspended|other|local]`

The tags in names are _not_ multiplied like `signin.count.status.[status].reason.[reason]`.

I think having reason without status will lead to confusion (which reasons are given for success and which one's for failing). So i would suggest naming the reason something like `failed:wrong-credentials`, to give the status explicitly. Or whatever other format is preferred.

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 4]
```diff
@@ -125,30 +154,89 @@ exports.signup = function(req, res) {
...
-      return res.status(403).send({
-        message: errorHandler.getErrorMessageByKey('suspended')
+      // Log the failure to signin
+      log('error', 'User signin failed. #3tfgbg-2', {
+        reason: 'Suspended user'
+      });
+
+      // Send signin failure to stats servers
+      statsObject.tags.status = 'failed';
+      statsObject.tags.reason = 'suspended';
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: same as line 179

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 5]
```diff
@@ -125,30 +154,89 @@ exports.signup = function(req, res) {
...
-      } else {
+        // Log the failure to signin
+        log('error', 'User signin failed. #3tfgbg-3', {
+          reason: 'Login error',
+          error: err
+        });
+
+        // Send signin failure to stats servers
+        statsObject.tags.status = 'failed';
+        statsObject.tags.reason = 'other';
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: same as line 179

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 6]
```diff
@@ -125,30 +154,89 @@ exports.signup = function(req, res) {
...
+          // Send error to the API
+          res.status(400).send(err);
+        });
+
+        return;
+      }
+
+      // Send signin success to stats servers
+      statsObject.tags.status = 'success';
+      statsObject.tags.reason = 'local';
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: same as line 179

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 7]
```diff
@@ -102,21 +99,53 @@ exports.signup = function(req, res) {
...
+  ], function(err, user) {
+
+    var statsObject = {
+      namespace: 'signup',
+      counts: {
+        count: 1
+      },
+      time: new Date(),
+      tags: {},
+      meta: {}
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: The `statsObject` looks good. `time`, `meta` and `tags` properties are optional, so `meta` can be removed and `time` depends on preference (time now vs. time of arrival to stathat).

File: modules/users/server/controllers/users/users.authentication.server.controller.js [hunk 8]
```diff
@@ -125,30 +154,89 @@ exports.signup = function(req, res) {
...
 exports.signin = function(req, res, next) {
+
+  var statsObject = {
+    namespace: 'signin',
+    counts: {
+      count: 1
+    },
+    time: new Date(),
+    tags: {},
+    meta: {}
```
Comment (mrkvon) [inline modules/users/server/controllers/users/users.authentication.server.controller.js]: the same as line 118

**Comment by mrkvon** [comment]
@simison the stats look fine to me. :+1: I added some recommendations for improvement to code review comments. It's optional to follow them.

**Comment by simison** [comment]
@mrkvon great! Thanks a bunch. I'll do these changes and merge then.

**Comment by simison** [comment]
@mrkvon Done!

So I'm basically pushing just `status` (so no separate `reason` anymore) as: 
- `success`
- `failed:suspended`
- `failed:wrong-credentials`
- `failed:other`

— good?

**Comment by mrkvon** [comment]
@simison Do we want to compare `success` and `failure` in general (in stathat)? If not, this is good.

Another possibility:
```js
{
  status: 'success'
}
```
and
```js
{
  status: 'failure',
  failureReason: ...reason...
}
```

But it doesn't matter, really. Choose any [colour](http://bikeshed.com/) you prefer. :slightly_smiling_face:

**Comment by simison** [comment]
@mrkvon these stats are mostly just to spot if something is broken.

**Pull request merged**